x/gamm/twap/logic.go

```diff
@@ -0,0 +1,149 @@
+	newSp1 := types.MustGetSpotPrice(k.ammkeeper, ctx, record.PoolId, record.Asset1Denom, record.Asset0Denom)
+
+	// set last spot price to be last price of this block. This is what will get used in interpolation.
+	newRecord.P0LastSpotPrice = newSp0
+	newRecord.P1LastSpotPrice = newSp1
+
+	return newRecord
+}
+
+// interpolate record returns a record, with updated accumulator values and time for provided newTime.
+//
+// pre-condition: newTime >= record.Time
+func recordWithUpdatedAccumulators(record types.TwapRecord, newTime time.Time) types.TwapRecord {
+	if record.Time.Equal(newTime) {
+		return record
+	}
+	newRecord := record
+	timeDelta := newTime.Sub(record.Time)
+	newRecord.Time = newTime
+
+	// record.LastSpotPrice is the last spot price from the block the record was created in,
+	// thus it is treated as the effective spot price until the new time.
+	// (As there was no change until at or after this time)
+	// TODO: Think about overflow
+	newRecord.P0ArithmeticTwapAccumulator = newRecord.P0ArithmeticTwapAccumulator.Add(
+		types.SpotPriceTimesDuration(record.P0LastSpotPrice, timeDelta))
+	newRecord.P1ArithmeticTwapAccumulator = newRecord.P1ArithmeticTwapAccumulator.Add(
+		types.SpotPriceTimesDuration(record.P1LastSpotPrice, timeDelta))
+	return newRecord
+}
+
+func (k Keeper) pruneOldTwaps(ctx sdk.Context) {
+	// TODO: Read this from parameter
+	lastAllowedTime := ctx.BlockTime().Add(-48 * time.Hour)
+	k.pruneRecordsBeforeTime(ctx, lastAllowedTime)
+}
+
+// getInterpolatedRecord returns a record for this pool, representing its accumulator state at time `t`.
+// This is achieved by getting the record `r` that is at, or immediately preceding in state time `t`.
+// To be clear: the record r s.t. `t - r.Time` is minimized AND `t >= r.Time`
+func (k Keeper) getInterpolatedRecord(ctx sdk.Context, poolId uint64, t time.Time, assetA, assetB string) (types.TwapRecord, error) {
+	if !(assetA > assetB) {
+		assetA, assetB = assetB, assetA
+	}
+	record, err := k.getRecordAtOrBeforeTime(ctx, poolId, t, assetA, assetB)
+	if err != nil {
+		return types.TwapRecord{}, err
+	}
+	record = interpolateRecord(record, t)
+	return record, nil
+}
+
+func (k Keeper) getMostRecentRecord(ctx sdk.Context, poolId uint64, assetA, assetB string) (types.TwapRecord, error) {
+	if !(assetA > assetB) {
+		assetA, assetB = assetB, assetA
+	}
+	record, err := k.getMostRecentRecordStoreRepresentation(ctx, poolId, assetA, assetB)
+	if err != nil {
+		return types.TwapRecord{}, err
+	}
+	record = interpolateRecord(record, ctx.BlockTime())
+	return record, nil
+}
+
+// interpolate record updates the record's accumulator values and time to the interpolate time.
+//
+// pre-condition: interpolateTime >= record.Time
+func interpolateRecord(record types.TwapRecord, interpolateTime time.Time) types.TwapRecord {
+	return recordWithUpdatedAccumulators(record, interpolateTime)
+}
+
+// TODO: Test math, test p0 vs p1 correctness
```

Comment:
I see a large test for this function (`TestComputeArithmeticTwap`). Does it cover it?

If not, would be great to keep track of this in an issue

Reply:
ah yeah it does, beyond comments in that function for rounding behavior and overflow

TY for catching